superset/assets/src/datasource/ChangeDatasourceModal.jsx

```diff
@@ -0,0 +1,170 @@
+    super(props);
+    this.state = {
+      loading: true,
+      datasources: null,
+    };
+    this.setSearchRef = this.setSearchRef.bind(this);
+    this.onEnterModal = this.onEnterModal.bind(this);
+    this.selectDatasource = this.selectDatasource.bind(this);
+    this.changeSearch = this.changeSearch.bind(this);
+  }
+
+  onEnterModal() {
+    if (this.searchRef) {
+      this.searchRef.focus();
+    }
+    if (!this.state.datasources) {
+      SupersetClient.get({
+        endpoint: '/superset/datasources/',
+      })
+        .then(({ json }) => {
+          const datasources = json.map(ds => ({
+            rawName: ds.name,
+            connection: ds.connection,
+            schema: ds.schema,
+            name: (
+              <a
+                href="#"
+                onClick={this.selectDatasource.bind(this, ds)}
+                className="datasource-link"
+              >
+                {ds.name}
+              </a>
+            ),
+            type: ds.type,
+          }));
+
+          this.setState({ loading: false, datasources });
+        })
+        .catch((response) => {
+          this.setState({ loading: false });
+          getClientErrorObject(response).then(({ error }) => {
+            this.props.addDangerToast(error.error || error.statusText || error);
+          });
+        });
+    }
+  }
+
+  setSearchRef(searchRef) {
+    this.searchRef = searchRef;
+  }
+
+  changeSearch(event) {
+    this.setState({ filter: event.target.value });
+  }
+
+  selectDatasource(datasource) {
+    SupersetClient.get({
+      endpoint: `/datasource/get/${datasource.type}/${datasource.id}`,
+    })
+      .then(({ json }) => {
+        this.props.onDatasourceSave(json);
+        this.props.onChange(datasource.uid);
+      })
+      .catch((response) => {
+        getClientErrorObject(response).then(({ error }) => {
+          this.props.addDangerToast(error.error || error.statusText || error);
+        });
+      });
+    this.props.onHide();
+  }
+
+  render() {
+    const { datasources, filter, loading } = this.state;
+    const { show, onHide } = this.props;
+
+    return (
+      <Modal
+        show={show}
+        onHide={onHide}
+        onEnter={this.onEnterModal}
+        onExit={this.setSearchRef}
+        bsSize="lg"
+      >
+        <Modal.Header closeButton>
+          <Modal.Title>{t('Select a datasource')}</Modal.Title>
+        </Modal.Header>
+        <Modal.Body>
+          <div>
+            <FormControl
+              id="formControlsText"
+              inputRef={(ref) => {
+                this.setSearchRef(ref);
+              }}
+              type="text"
+              bsSize="sm"
+              value={filter}
+              placeholder={t('Search / Filter')}
+              onChange={this.changeSearch}
+            />
+          </div>
+          {loading && <Loading />}
+          {datasources && (
+            <Table
+              columns={['name', 'type', 'schema', 'connection', 'creator']}
```

Comment:
Can make this a constant so in the case that `Table` does shallow comparison, it won't detect columns array as new value every time.